File: app/src/main/java/com/example/alockwood/windowinsetsplayground/HomeFragment.java
package com.example.alockwood.windowinsetsplayground;

import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

public class HomeFragment extends BaseFragment implements OnMapReadyCallback {

  private static final String MAP_VIEW_BUNDLE_KEY = "map_view_bundle_key";

  private MapView mapView;

  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    final View view = inflater.inflate(R.layout.fragment_home, container, false);
    mapView = view.findViewById(R.id.map_view);
    return view;
  }

  @Override
  public void onActivityCreated(@Nullable Bundle savedInstanceState) {
    super.onActivityCreated(savedInstanceState);

    Bundle mapViewBundle = null;
    if (savedInstanceState != null) {
      mapViewBundle = savedInstanceState.getBundle(MAP_VIEW_BUNDLE_KEY);
    }
    mapView.onCreate(mapViewBundle);
    mapView.getMapAsync(this);
  }

  @Override
  public void onSaveInstanceState(@NonNull Bundle outState) {
    super.onSaveInstanceState(outState);
    Bundle mapViewBundle = outState.getBundle(MAP_VIEW_BUNDLE_KEY);
    if (mapViewBundle == null) {
      mapViewBundle = new Bundle();
      outState.putBundle(MAP_VIEW_BUNDLE_KEY, mapViewBundle);
    }
    mapView.onSaveInstanceState(mapViewBundle);
  }

  @Override
  public void onStart() {
    super.onStart();
    mapView.onStart();
  }

  @Override
  public void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override
  public void onPause() {
    super.onPause();
    mapView.onPause();
  }

  @Override
  public void onStop() {
    super.onStop();
    mapView.onStop();
  }

  @Override
  public void onDestroy() {
    super.onDestroy();
    mapView.onDestroy();
  }

  @Override
  public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }

  @Override
  public void onMapReady(GoogleMap googleMap) {
    googleMap.addMarker(new MarkerOptions().position(new LatLng(0, 0)).title("Marker"));
  }

  @Override
  protected int getStatusBarBackground() {
    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
        ? R.color.design_core_ui_white_alpha70
        : R.color.design_core_ui_black_alpha40;
  }

  @Override
  protected boolean shouldUseLightStatusBar() {
    return Build.VERSION.SDK_INT >= Build.VERSION_CODES.M;
  }

  @Override
  protected boolean shouldDrawContentUnderStatusBar() {
    return true;
  }
}

File: app/src/main/java/com/example/alockwood/windowinsetsplayground/HomeFragment.kt
package com.example.alockwood.windowinsetsplayground

import android.os.Build
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions

class HomeFragment : Fragment(), OnMapReadyCallback {

    private lateinit var mapView: MapView

    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_home, container, false)
        mapView = view.findViewById(R.id.map_view)

        val statusBarBackgroundView: View = view.findViewById(R.id.status_bar_background_view)
        statusBarBackgroundView.setBackgroundResource(
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    R.color.design_core_ui_white_alpha70
                } else {
                    R.color.design_core_ui_black_alpha40
                })

        view.setOnApplyWindowInsetsListener { v, insets ->
            val mapLayoutParams = mapView.layoutParams as ViewGroup.MarginLayoutParams
            mapLayoutParams.leftMargin = insets.systemWindowInsetLeft
            mapLayoutParams.rightMargin = insets.systemWindowInsetRight
            mapLayoutParams.bottomMargin = insets.systemWindowInsetBottom
            mapView.layoutParams = mapLayoutParams

            val statusBarLayoutParams = statusBarBackgroundView.layoutParams
            statusBarLayoutParams.height = insets.systemWindowInsetTop
            statusBarBackgroundView.layoutParams = statusBarLayoutParams

            insets.consumeSystemWindowInsets()
        }
        return view
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)

        var mapViewBundle: Bundle? = null
        if (savedInstanceState != null) {
            mapViewBundle = savedInstanceState.getBundle(MAP_VIEW_BUNDLE_KEY)
        }
        mapView.onCreate(mapViewBundle)
        mapView.getMapAsync(this)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        var mapViewBundle = outState.getBundle(MAP_VIEW_BUNDLE_KEY)
        if (mapViewBundle == null) {
            mapViewBundle = Bundle()
            outState.putBundle(MAP_VIEW_BUNDLE_KEY, mapViewBundle)
        }
        mapView.onSaveInstanceState(mapViewBundle)
    }

    override fun onStart() {
        super.onStart()
        mapView.onStart()
    }

    override fun onResume() {
        super.onResume()
        mapView.onResume()
    }

    override fun onPause() {
        super.onPause()
        mapView.onPause()
    }

    override fun onStop() {
        super.onStop()
        mapView.onStop()
    }

    override fun onDestroy() {
        super.onDestroy()
        mapView.onDestroy()
    }

    override fun onLowMemory() {
        super.onLowMemory()
        mapView.onLowMemory()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        googleMap.addMarker(MarkerOptions().position(LatLng(0.0, 0.0)).title("Marker"))
    }

    companion object {

        private val MAP_VIEW_BUNDLE_KEY = "map_view_bundle_key"
    }
}
